accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+`mapEntry` elements. *(We could avoid this prose by duplicating the above rules
+for lists and removing `mapEntry`, but this is simpler.)*
+
+The existing rule that an expression statement cannot start with `{` now covers
+both map and set literals.
+
+## Static Semantics
+
+### Disambiguating sets and maps
+
+Because sets and maps both use curly braces as delimiters, you can create
+collection literals that are not obviously either a set or a map, like:
+
+```dart
+var a = {};
+var b = {...other};
+```
+
+When possible, we first use syntax and the surrounding context type to
+disambiguate between a map and set:
+
+Let *e* be a `setOrMapLiteral`.
+
+1.  If *e* has `typeArguments` then:
+
+    *   If there is exactly one type argument `T`, then *e* is a set literal
+        with static type `Set<T>`.
+
+    *   If there are exactly two type arguments `K` and `V`, then *e* is a map
+        literal with static type `Map<K, V>`.
+
+    *   Otherwise (three or more type arguments), report a compile-time error.
+
+1.  Else, if *e* has a context `C`, and the base type of `C` is `Cbase` (that
+    is, `Cbase` is `C` with all wrapping `FutureOr`s removed), and `Cbase` is
+    not `?`, and `S` is the greatest closure of `Cbase` then:
+
+    *   If `S` is a subtype of `Iterable<Object>` and `S` is not a subtype of
+        `Map<Object, Object>`, then *e* is a set literal.
+
+    *   If `S` is a subtype of `Map<Object, Object>` and `S` is not a subtype of
+        `Iterable<Object>` then *e* is a map literal.
+
+1.  Else, if *leaf elements* is not empty, then:
+
+    *   If *leaf elements* has at least one `expressionElement` and no
+        `mapEntry` elements, then *e* is a set literal with unknown static type.
+
+    *   If *leaf elements* has at least one `mapEntry` and no
+        `expressionElement` elements, then *e* is a map literal with unknown
+        static type.
+
+    *   If *leaf elements* has at least one `mapEntry` and at least one
+        `expressionElement`, report a compile-time error.
+
+    *In other words, at least one key-value pair anywhere in the collection
+    forces it to be a map, and a bare expression forces it to be a set. Having
+    both is an error.*
+
+If *e* has no `typeArguments` and no context type, and no `elements`, then *e*
+is treated as a map literal with unknown static type. *In other words, an empty
+`{}` is a map unless we have a context that indicates otherwise.*
+
+At this point, there are three states we could be in:
+
+*   *e* contains a compile-time error. In that case, we don't proceed.
+
+*   *e* is disambiguated to definitely be either a map or set literal.
+
+*   *e* is still ambiguous. This can only happen when *e* is non-empty but
+    contains only `spreadElement`s. In this case, the disambiguation will happen
+    during type inference, defined below.
+
+### Type inference
+
+#### Maps and sets
+
+Inference and set/map disambiguation are done concurrently. We perform inference
```

Comment:
OK, moved the sentence. I'm a little unclear on how specific we need to be around "phases" of this process. I'm trying to minimize the surgery I do on this part of the proposal since it's mostly coming from Leaf and I don't know if there are important pieces of the ordering that I'm oblivious to.